### PR TITLE
MAINT: QMCEngine d input validation

### PR DESCRIPTION
**Pull request by tupui**
As pointed out in #13319, `d` is not checked when instantiating `QMCEngine`.

### CODE REVIEW

File: scipy/stats/_qmc.py [hunk 1]
```diff
@@ -968,11 +971,11 @@ class Sobol(QMCEngine):
     MAXBIT = _MAXBIT
 
     def __init__(self, d, scramble=True, seed=None):
+        super().__init__(d=d, seed=seed)
```
Comment (mdhaber) [inline scipy/stats/_qmc.py:974]: Is this only needed in `Sobol`?

File: scipy/stats/tests/test_qmc.py [hunk 2]
```diff
@@ -574,9 +578,10 @@ def test_random_base2(self):
             engine.random_base2(2)
 
     def test_raise(self):
+        seed = np.random.RandomState(12345)
         with pytest.raises(ValueError, match=r"Maximum supported "
                                              r"dimensionality"):
-            qmc.Sobol(qmc.Sobol.MAXDIM + 1)
+            qmc.Sobol(qmc.Sobol.MAXDIM + 1, seed=seed)
```
Comment (mdhaber) [inline scipy/stats/tests/test_qmc.py:584]: This is unrelated, right? (That's ok, since this PR is so simple - just checking.)

File: scipy/stats/_qmc.py [hunk 3]
```diff
@@ -567,6 +567,9 @@ class QMCEngine(ABC):
 
     @abstractmethod
     def __init__(self, d, seed=None):
+        if not isinstance(d, int):
```
Comment (mdhaber) [inline scipy/stats/_qmc.py:570]:
```suggestion
        if not np.issubdtype(type(d), np.integer):
```
I think this would be a little more flexible, allowing e.g. `np.int64`.

File: scipy/stats/_qmc.py (hunk 1 above)
Comment (tupui) [reply scipy/stats/_qmc.py:974]: The others were doing this already at the beginning of the init so it's fine. I would need to do some adjustments for the multivariate ones now that I re-read the code a bit, but it's unrelated to this.

File: scipy/stats/tests/test_qmc.py (hunk 2 above)
Comment (tupui) [reply scipy/stats/tests/test_qmc.py:584]: It was not needed before as the init happened before this validation. Now that `super` is called first, it's needed.

File: scipy/stats/_qmc.py (hunk 3 above)
Comment (tupui) [reply scipy/stats/_qmc.py]: Ah ok sure, thanks!

**Comment by mdhaber** [comment]
LGTM thanks @tupui!

**Comment by tupui** [comment]
Thanks for the quick review Matt.
